stdlib/public/core/Span/RawSpan.swift

```diff
@@ -0,0 +1,702 @@
+  @usableFromInline @inline(__always)
+  @lifetime(immortal)
+  init(
+    _unchecked pointer: UnsafeRawPointer?,
+    byteCount: Int
+  ) {
+    _pointer = pointer
+    _count = byteCount
+  }
+}
+
+@_disallowFeatureSuppression(NonescapableTypes)
+@available(SwiftStdlib 9999, *)
+extension RawSpan: @unchecked Sendable {}
+
+@_disallowFeatureSuppression(NonescapableTypes)
+@available(SwiftStdlib 9999, *)
+extension RawSpan {
+
+  /// Unsafely create a `RawSpan` over initialized memory.
+  ///
+  /// The memory in `buffer` must be owned by the instance `owner`,
+  /// meaning that as long as `owner` is alive the memory will remain valid.
+  ///
+  /// - Parameters:
+  ///   - buffer: an `UnsafeRawBufferPointer` to initialized memory.
+  ///   - owner: a binding whose lifetime must exceed that of
+  ///            the newly created `RawSpan`.
+  @_disallowFeatureSuppression(NonescapableTypes)
+  @_alwaysEmitIntoClient
+  @lifetime(immortal)
+  public init(
+    _unsafeBytes buffer: UnsafeRawBufferPointer
+  ) {
+    self.init(
+      _unchecked: buffer.baseAddress, byteCount: buffer.count
+    )
+  }
+
+  @_disallowFeatureSuppression(NonescapableTypes)
+  @_alwaysEmitIntoClient
+  @lifetime(immortal)
+  public init(
+    _unsafeBytes buffer: Slice<UnsafeRawBufferPointer>
+  ) {
+    self.init(_unsafeBytes: UnsafeRawBufferPointer(rebasing: buffer))
+  }
+
+  /// Unsafely create a `RawSpan` over initialized memory.
+  ///
+  /// The memory in `buffer` must be owned by the instance `owner`,
+  /// meaning that as long as `owner` is alive the memory will remain valid.
+  ///
+  /// - Parameters:
+  ///   - buffer: an `UnsafeMutableRawBufferPointer` to initialized memory.
+  ///   - owner: a binding whose lifetime must exceed that of
+  ///            the newly created `RawSpan`.
+  @_disallowFeatureSuppression(NonescapableTypes)
+  @_alwaysEmitIntoClient
+  @lifetime(immortal)
+  public init(
+    _unsafeBytes buffer: UnsafeMutableRawBufferPointer
+  ) {
+    self.init(_unsafeBytes: UnsafeRawBufferPointer(buffer))
+  }
+
+  @_disallowFeatureSuppression(NonescapableTypes)
+  @_alwaysEmitIntoClient
+  @lifetime(immortal)
+  public init(
+    _unsafeBytes buffer: Slice<UnsafeMutableRawBufferPointer>
+  ) {
+    self.init(_unsafeBytes: UnsafeRawBufferPointer(rebasing: buffer))
+  }
+
+  /// Unsafely create a `RawSpan` over initialized memory.
+  ///
+  /// The memory over `count` bytes starting at
+  /// `pointer` must be owned by the instance `owner`,
+  /// meaning that as long as `owner` is alive the memory will remain valid.
+  ///
+  /// - Parameters:
+  ///   - pointer: a pointer to the first initialized byte.
+  ///   - byteCount: the number of initialized bytes in the span.
+  ///   - owner: a binding whose lifetime must exceed that of
+  ///            the newly created `RawSpan`.
+  @_disallowFeatureSuppression(NonescapableTypes)
+  @_alwaysEmitIntoClient
+  @lifetime(immortal)
+  public init(
+    _unsafeStart pointer: UnsafeRawPointer,
```

Comment:
Should this parameter have optional type?

Reply:
The `Unsafe[*]BufferPointer`-taking initializers are the way to supply an optional pointer; it cuts down a bit on the number of required precondition checks, but that could be an unnecessary optimization.

Reply:
The precondition check we'd have to add is this, right?

```swift
_precondition(count == 0 || pointer != nil)
```

So if I'm coming in from a C entry point that has separate pointer + count, where the pointer itself is generally going to be optional, the right way to do this is to go through UnsafeBufferPointer. Now, the UnsafeBufferPointer initializer only has a `_debugPrecondition` assertion for this condition, so in release builds the invariant doesn't hold for UnsafeBufferPointer. It looks like Span's init from an UnsafeBufferPointer doesn't validate this condition, either.

Reply:
The stdlib's approach has been to assume that e.g. a `Range` or `UnsafeBufferPointer` is well-formed when received as a parameter, even though it is _possible_ to create ill-formed ones. This may be a situation vigilance beyond the usual is indicated, but I'm not certain. I will think about it!

Reply:
I do expect the optimizer to eliminate the precondition if the function takes an optional but gets passed a non-optional pointer (if the `||` autoclosure interferes with that, then we can swap the order). If the worry is that this function may not get inlined, then we can force inline it.